### PR TITLE
Fixed correctly flushing send buffers on closing a socket.

### PR DESCRIPTION
**Pull request by ethouris**
This removes the previous trial implementation of preventing too early closing the socket and an extra mutex. This was experimental and blocked. The old implementation has been removed, the new one applies only closing when the send buffer is empty, updating the checks on `waitForEvent`.

### CODE REVIEW

File: srtcore/api.cpp [hunk 1]
```diff
@@ -953,53 +955,72 @@ int CUDTUnited::close(const SRTSOCKET u)
...
+           // The socket is either in m_ClosedSockets or is already gone.
 
-           // Do a 0.20s rolling, just for safety, when we missed the signal.
-           timeval now;
-           timespec timeout;
-           gettimeofday(&now, 0);
-           timeout.tv_sec = now.tv_sec;
-           timeout.tv_nsec = (now.tv_usec + 200000) * 1000;
+           // Done the other way, but still done. You can stop waiting.
+           if ( m_ClosedSockets.count(id) == 0 || !s->m_pUDT->m_bOpened )
```
Comment (rndi) [inline srtcore/api.cpp]: I think you need to take m_ControlLock before accessing m_ClosedSockets. Bizarrely enough the code you are replacing was not taking that lock either and also was also calling pthread_cond_timedwait() with lock not taken which is undefined. Unless I am missing something obvious of course, - it definitely feels like it.

File: srtcore/api.cpp [hunk 2]
```diff
@@ -953,53 +955,72 @@ int CUDTUnited::close(const SRTSOCKET u)
...
+   /*
+      This code is PUT ASIDE for now.
+      Most likely this will be never required.
+      It had to hold the closing activity until the time when the receiver buffer is depleted.
+      However the closing of the socket should only happen when the receiver has received
+      an information about that the reading is no longer possible (error report from recv/recvfile).
+      When this happens, the receiver buffer is definitely depleted already and there's no need to check
+      anything.
+
+      Should there appear any other conditions in future under which the closing process should be
```
Comment (rndi) [inline srtcore/api.cpp]: What happens if close() is called from a different thread while read is still a progress? I feel like we may be changing API behavior here.

File: srtcore/api.cpp (hunk 1 above)
Comment (ethouris) [reply srtcore/api.cpp]: The code that was there before was experimental and not enabled by default. There was a new mutex and cv also introduced there; I removed them as they were not necessary. I'll take a look at race issues.

File: srtcore/api.cpp (hunk 2 above)
Comment (ethouris) [reply srtcore/api.cpp]: This is not subject of this fix here. This fix only takes care of the problem of closing the socket too early, when the sending buffer is not yet empty (which caused that the data already scheduled for sending by `srt_sendmsg` had no chance to be physically sent over the network and therefore completely lost).

In a case of reading the standard procedure is that the reading is interrupted and reports error. In this case this is done by signaling appropriate cv, which causes to exit the data-waiting loop (and set the event for epoll), but without delivering any data.

**Pull request merged**